commit 48b462c31bdcb7156a2d7a34d5addb8e9421e42e
Author: Marcin Pawlowski <mpawlowski@fb.com>
Date:   Wed Aug 10 16:01:06 2016 -0700

    yummy:patchX:core--introduce-UseRootFileSystemNamespace-option.patch

core: introduce UseRootFileSystemNamespace option

    
    Allow user to specify that unit has to run in init's files system
    namespace. Previously it was possible, iff none of the file system
    namespace related option was set (but documentation gave no such
    warranty). I wanted to have a formal and documented option which
    forces such behavior. This option is incompatible with: PrivateTmp=,
    PrivateDevices=, ProtectSystem=, ProtectHome=, ReadOnlyPaths=,
    InaccessiblePaths=, ReadWritePaths=, MountFlags=, meaning it
    cannot be set to true, if any of the listed options has non default
    value.

diff --git a/man/systemd.exec.xml b/man/systemd.exec.xml
index 41ae6e7..82a3b01 100644
--- a/man/systemd.exec.xml
+++ b/man/systemd.exec.xml
@@ -1013,6 +1013,26 @@
       </varlistentry>
 
       <varlistentry>
+        <term><varname>UseRootFileSystemNamespace=</varname></term>
+
+        <listitem><para>Takes a boolean argument (default is false).
+        Forces unit's process to start in the same file system namespace
+        as init (the namespace of systemd). Setting this option to
+        <literal>true</literal> disables all file system related options,
+        namely:
+        (<varname>MountFlags=</varname>,
+        <varname>PrivateTmp=</varname>,
+        <varname>PrivateDevices=</varname>,
+        <varname>ProtectSystem=</varname>,
+        <varname>ProtectHome=</varname>,
+        <varname>ReadOnlyPaths=</varname>,
+        <varname>InaccessiblePaths=</varname> and
+        <varname>ReadWritePaths=</varname>). If any of these options
+        is set, the unit is deemed incorrectly configured.
+        </para></listitem>
+      </varlistentry>
+
+      <varlistentry>
         <term><varname>MountFlags=</varname></term>
 
         <listitem><para>Takes a mount propagation flag:
diff --git a/src/core/execute.c b/src/core/execute.c
index 7c178b9..b417630 100644
--- a/src/core/execute.c
+++ b/src/core/execute.c
@@ -1913,7 +1913,9 @@ static int exec_child(
                 }
         }
 
-        needs_mount_namespace = exec_needs_mount_namespace(context, params, runtime);
+        needs_mount_namespace =
+                !context->use_root_file_system_namespace &&
+                exec_needs_mount_namespace(context, params, runtime);
 
         if (needs_mount_namespace) {
                 char *tmp = NULL, *var = NULL;
diff --git a/src/core/execute.h b/src/core/execute.h
index 189c4d0..1681bd5 100644
--- a/src/core/execute.h
+++ b/src/core/execute.h
@@ -171,6 +171,7 @@ struct ExecContext {
         bool private_devices;
         ProtectSystem protect_system;
         ProtectHome protect_home;
+        bool use_root_file_system_namespace;
 
         bool no_new_privileges;
 
diff --git a/src/core/load-fragment-gperf.gperf.m4 b/src/core/load-fragment-gperf.gperf.m4
index 125c760..2188f1f 100644
--- a/src/core/load-fragment-gperf.gperf.m4
+++ b/src/core/load-fragment-gperf.gperf.m4
@@ -89,6 +89,7 @@ $1.InaccessiblePaths,            config_parse_namespace_path_strv,   0,
 $1.PrivateTmp,                   config_parse_bool,                  0,                             offsetof($1, exec_context.private_tmp)
 $1.PrivateNetwork,               config_parse_bool,                  0,                             offsetof($1, exec_context.private_network)
 $1.PrivateDevices,               config_parse_bool,                  0,                             offsetof($1, exec_context.private_devices)
+$1.UseRootFileSystemNamespace,   config_parse_bool,                  0,                             offsetof($1, exec_context.use_root_file_system_namespace)   
 $1.ProtectSystem,                config_parse_protect_system,        0,                             offsetof($1, exec_context)
 $1.ProtectHome,                  config_parse_protect_home,          0,                             offsetof($1, exec_context)
 $1.MountFlags,                   config_parse_exec_mount_flags,      0,                             offsetof($1, exec_context)
diff --git a/src/core/load-fragment.c b/src/core/load-fragment.c
index aa19e42..5878ea0 100644
--- a/src/core/load-fragment.c
+++ b/src/core/load-fragment.c
@@ -3696,6 +3696,25 @@ int config_parse_protect_system(
         return 0;
 }
 
+static bool verify_mount_ns_config(const ExecContext* context) {
+    assert(context);
+
+    if (!context->use_root_file_system_namespace)
+            return true;
+
+    if (context->private_tmp ||
+        context->private_devices ||
+        context->protect_system != PROTECT_SYSTEM_NO ||
+        context->protect_home != PROTECT_HOME_NO ||
+        context->mount_flags != 0 ||
+        !strv_isempty(context->read_write_paths) ||
+        !strv_isempty(context->read_only_paths) ||
+        !strv_isempty(context->inaccessible_paths))
+            return false;
+
+    return true;
+}
+
 #define FOLLOW_MAX 8
 
 static int open_follow(char **filename, FILE **_f, Set *names, char **_final) {
@@ -3924,6 +3943,16 @@ static int load_from_path(Unit *u, const char *path) {
                                  false, true, false, u);
                 if (r < 0)
                         return r;
+
+
+                if (UNIT_HAS_EXEC_CONTEXT(u) &&
+                    !verify_mount_ns_config(unit_get_exec_context(u))) {
+                        log_unit_warning(u, "In unit %s, option UseRootFileSystemNamespace=true is incompatible with "
+                                            "file system namespace related options: PrivateTmp, PrivateDevices, "
+                                            "ProtectSystem, ProtectHome, ReadOnlyPaths, InaccessiblePaths and   "
+                                            "ReadWritePaths", u->id);
+                        return -EINVAL;
+                }
         }
 
         free(u->fragment_path);
